src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala

```diff
@@ -63,26 +77,51 @@ object ErgoBoxCandidate {
       val propBytes = obj.propositionBytes
       val (regBytes, regNum) = collectRegister(obj, Array.emptyByteArray, 0: Byte)
 
-      Longs.toByteArray(obj.value) ++ propBytes ++ (regNum +: regBytes)
+      val tokensCount = obj.additionalTokens.size.toByte
+      val tokenBytes = if (obj.additionalTokens.nonEmpty) {
+        obj.additionalTokens.map { case (id, amount) =>
+          id ++ Longs.toByteArray(amount)
+        }.reduce(_ ++ _)
+      } else {
+        Array.emptyByteArray
+      }
+
+      Longs.toByteArray(obj.value) ++ propBytes ++ (tokensCount +: tokenBytes) ++ (regNum +: regBytes)
```

Comment:
@kushti use Writer to serialize data in a byte array and avoid allocation of intermediate arrays. Also code will be shorter and easier to understand.

Reply:
@ergomorphic @kushti FWIW I've switched the `ErgoBoxCandidate` serializer to reader/writer API here - https://github.com/ScorexFoundation/sigmastate-interpreter/blob/e35a8eea77939443bfcff7a2f62b7648ee18992d/src/main/scala/org/ergoplatform/ErgoBoxCandidate.scala#L74